### PR TITLE
Fixed issue 20422 - confusing error "new can only create structs, dynamic arrays or class objects, not int[]'s"

### PR DESCRIPTION
**Pull request by VladNastase**

### CODE REVIEW

**Review by thewilsonator — APPROVED** [review]

**Comment by thewilsonator** [comment]
Please squash  the commits and title the commit message as "Fix issue 20422 - confusing error "new can only create structs, dynamic arrays or class objects, not int[]'s" and this should be good to go.

File: src/dmd/expressionsem.d [hunk 1]
```diff
@@ -3775,37 +3775,47 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
...
+                    (*exp.arguments)[i] = arg;
+                    tb = (cast(TypeDArray)tb).next.toBasetype();
                 }
-                (*exp.arguments)[i] = arg;
-                tb = (cast(TypeDArray)tb).next.toBasetype();
+            } 
+            else 
+            {
+                // https://issues.dlang.org/show_bug.cgi?id=20422
+                // Without this check the compiler would give a missleading error
```
Comment (Geod24) [inline src/dmd/expressionsem.d:3816]: misleading

**Review by Geod24 — COMMENTED** [review]
I'd recommend inverting the condition:
```D
if (!narg)
{
    // ...
    return setError();
}
// Code without else
```

This makes the code more readable, as you get the error cases "out of the way".

**Comment by VladNastase** [comment]
> I'd recommend inverting the condition:
> 
> ```d
> if (!narg)
> {
>     // ...
>     return setError();
> }
> // Code without else
> ```
> 
> This makes the code more readable, as you get the error cases "out of the way".

Done! Thanks for the suggestion!

**Review by RazvanN7 — APPROVED** [review]

**Pull request merged**